File: python/vayu_sdk/clients/vayu_client.py
from openapi.api_client import ApiClient
from openapi.configuration import Configuration
from openapi.rest import RESTClientObject
from vayu_sdk.clients.authenticator import VayuAuthenticator


class VayuClient:
    def __init__(self, api_key: str, host: str = None):
        self._host = host
        self.auth = VayuAuthenticator(api_key, host)

    @property
    def access_token(self):
        return self.auth.access_token

    @property
    def is_logged_in(self):
        return self.auth.is_authenticated

    @property
    def client(self):
        configuration = Configuration(host=self._host)
        client = ApiClient(configuration)
        client.rest_client = AuthenticatedRESTClient(self.auth, configuration)
        return client

    def login(self):
        """Deprecated: Authentication is now handled automatically. You can remove this call."""
        self.auth.authenticate()

    def validate_logged_in(self):
        """Deprecated: Authentication is now handled automatically."""
        pass


class AuthenticatedRESTClient(RESTClientObject):
    """Thin REST client that delegates auth to VayuAuthenticator and retries on 401."""

    def __init__(self, authenticator: VayuAuthenticator, configuration: Configuration):
        super().__init__(configuration)
        self._auth = authenticator

    def request(self, method, url, headers=None, **kwargs):
        is_login_request = "/login" in url

        if not is_login_request:
            token = self._auth.ensure_valid_token()
            if headers is None:
                headers = {}
            headers["Authorization"] = f"Bearer {token}"

        response = super().request(method, url, headers=headers, **kwargs)

        # 401 retry: re-authenticate and retry once
        if not is_login_request and response.status == 401:
            self._auth.authenticate()
            headers["Authorization"] = f"Bearer {self._auth.access_token}"
            response = super().request(method, url, headers=headers, **kwargs)

        return response


__all__ = ["VayuClient"]

File: go/client/vayu_client.go
package client

import (
	"net/http"

	"github.com/weft-finance/vayu-go/openapi"
)

type VayuClient struct {
	auth   *Authenticator
	Client *openapi.APIClient
}

func NewVayuClient(APIKey string) *VayuClient {
	apiClient := openapi.NewAPIClient(openapi.NewConfiguration())

	cfg := apiClient.GetConfig()
	cfg.UserAgent = "VayuSDK/go"

	auth := NewAuthenticator(APIKey, apiClient)

	vc := &VayuClient{
		auth:   auth,
		Client: apiClient,
	}

	vc.installRoundTripper()

	return vc
}

func (vc *VayuClient) SetCustomHost(host string) {
	cfg := vc.Client.GetConfig()
	cfg.Servers = openapi.ServerConfigurations{
		{
			URL:         host,
			Description: "Custom server",
		},
	}
}

// Deprecated: Authentication is now handled automatically. You can remove this call.
func (vc *VayuClient) Login() error {
	return vc.auth.Authenticate()
}

func (vc *VayuClient) IsLoggedIn() bool {
	return vc.auth.IsAuthenticated()
}

// Deprecated: Authentication is now handled automatically.
func (vc *VayuClient) ValidateLoggedIn() error {
	return nil
}

func (vc *VayuClient) installRoundTripper() {
	cfg := vc.Client.GetConfig()

	if cfg.HTTPClient.Transport == nil {
		cfg.HTTPClient.Transport = http.DefaultTransport
	}

	cfg.HTTPClient.Transport = &vayuRoundTripper{
		rt:   cfg.HTTPClient.Transport,
		auth: vc.auth,
	}
}
